Catch zero standard deviation in data scaling logic.

troiganto opened the pull request:
Closes #99.

Contrary to the suggestion given by the @charlie5, I decided to put the check to where the standard deviation is calculated, not where the scaling happens. This way, we handle both scaling and descaling in one go.

I set the standard deviation from zero to one so that there is still *some* sort of scaling happening. I guess another option would be to just return some form of error on zero.

andersfylling commented on src/fann_train_data.c:
```diff
@@ -1186,8 +1186,10 @@ FANN_EXTERNAL void FANN_API fann_descale_train( struct fann *ann, struct fann_tr
+			ann->scale_deviation_##where[ cur_neuron ] == 0.0f ? 1.0 :                                  \
+				sqrtf( ann->scale_deviation_##where[ cur_neuron ] / (float)data->num_data ); 			\
 	/* Calculate factor: (new_max-new_min)/(old_max(1)-old_min(-1)) */									\
 	/* Looks like we dont need whole array of factors? */												\
```
Isn't it better to set it to 0.0f instead of 1.0, since the developer expects the input to be 0.0f?

I'm not entirely certain if this is the only place you get the divided by zero issue, but if it is, then only dividing in the "else block", as youve done, should be enough?

troiganto replied:
The division by `data->num_data` here is actually unproblematic, the functions `fann_set_input_scaling_params()` and `fann_set_output_scaling_params()` already guard against this case. (and set mean and deviation to the default values 0.0 and 1.0 resp.)

The problem described in issue #99 happens in `fann_scale_train()` and `fann_descale_train()`. The former divides by the standard deviation, the latter mulitplies by it. This can happen if the values of one input or output column are all the same and we currently don't guard against that. Good catch though, I'll expand the comment above this line to point to the problematic functions.

troiganto replied:
Not sure if it's a good idea, but to make it more obvious what's going on, we could also put the check into a separate statement. Something like this:

```c
for( cur_neuron = 0; cur_neuron < ann->num_##where##put; cur_neuron++ ) {
  ann->scale_deviation_##where[ cur_neuron ] =
    sqrtf( ann->scale_deviation_##where[ cur_neuron ] / (float)data->num_data );
  if( ann->scale_deviation_##where[ cur_neuron ] == 0.0f ) {
    /* explanatory comment here */
    ann->scale_deviation_##where[ cur_neuron ] = 1.0f;
  }
}
```
But using the ternary operator as in the original commit seemed more ~efficient~ to me.

andersfylling replied:
when reviewing the issue I see that he ignored every neuron with a val of 0.0f. As they caused an issue:
```
if(ann->scale_deviation_in[ cur_neuron ] != 0.0)
```

But now those are changed to 1.0f later. My concern is what happens to later code, that has been developed with it being 0.0f in mind?

troiganto replied:
My approach differs from the one suggested in issue #99. Their solution was: if the standard deviation of an input is zero, don't scale it at all. My solution still scales such an input, but with an (admittedly arbitrarily chosen) standard deviation of 1.0.

The advantage of the original solution is that it's certainly less surprising. The advantage of mine is that even in the case of the input being constant, it is scaled into the target range `[new_min; new_max]`.

As far as I can tell, we're pretty much free to choose either option, since the docs explicitly don't guarantee that the input will be in-range after scaling. I'm not having a strong opinion either and the change to the solution originally suggested isn't too complicated.

Would you prefer we solve this case as originally suggested by not scaling these degenerate inputs at all?

andersfylling replied:
I'm leaning towards to ignore cases with 0.0f:
```
ann->scale_deviation_##where[ cur_neuron ] == 0.0f ? 0.0f : sqrtf( ann->scale_deviation_##where[ cur_neuron ] / (float)data->num_data );
```

andersfylling replied:
Thoughts? @Bukka

charlie5 replied:
Would it be sensible to disable an input neuron with 0 deviation ?

bukka replied:
I would be more in favor of keeping that as `0.0f` and ignore those neurons in scaling. The thing is that if you already  saved such `ann` by `fann_save` (let's say before you updated the library with the new fix changing that to 1.0f) and then loaded `ann` by `fann_load` after you did the lib update, it will still have value 0.0f so we would still need to add those checks to `fann_descale` and `fann_scale`...

Also It seems slightly safer in case some code out there would be using `scale_deviation_[in|out]` explicitly and relied on it to be a zero but that's quite unlikely IMHO.

troiganto commented:
Hi @andersfylling, it's been a while since this PR was last updated. Would you mind having another look and merging it if there're no further issues? This is my first PR for libfann, so I wouldn't be comfortable merging it myself. :slightly_smiling_face:

troiganto commented:
Alright, I've pushed a revised history that implements the suggested changes. Instead of changing the standard deviation from `0.0` to `1.0` when setting the scaling params, the scale/descale functions now check for a standard deviation of `0.0` and ignore all inputs/outputs where this is the case.

Remaining issues:
* Do we have to consider the case where `ann->scale_factor_*` is zero? This is only the case if `new_max == new_min`.
* I'm ignoring degenerate inputs/outputs both when scaling and when descaling. Is this okay or should we ignore them only when we'd actually *divide* by zero?
* One of the CI jobs fails, even after restarting. Looking at the details, it looks like it might be related to the fact that I deleted the original two commits. Do you guys know a way to fix this issue?

One more note: When pushing my new changes, I accidentally pushed to the upstream master branch instead of my fork's. I noticed the mistake within a few seconds and did a hard reset back to [this commit](https://github.com/libfann/fann/commit/b362b4bf8440f16ecf408ed6ab2a788bdea5c65f). I've also changed my setup so that the automatic push now really goes to my fork instead of the main repo. I'm sorry if I broke anything.

troiganto commented:
Okay, the merge error was definitely caused by me messing up the push to my fork. Unfortunately, while trying to fix it, I removed the merge of #103 and merged this PR instead, making everything worse.

@bukka, I'm not entirely sure what the best way forward is from here. Can you unmerge this PR, and somehow add add the lost commit back to the history? It's still available by its [commit hash](https://github.com/libfann/fann/commit/2e48bb14f61155423565c6ba77f1b8b596ebd04d) and I've made a backup [on my fork](https://github.com/troiganto/fann/tree/troiganto/pr-103), just to be safe. But I'd rather not break even more than I already have.

Again, I'm very sorry for the caused inconvenience.

bukka commented:
No worries :) Actually not sure how you managed to do it as it should be protected against forced push (at lest it didn't work for me). So I did at least the quick merge and push so it is back. It might break pull for some users thought as the history changed. Anyway good thing to do before the merge is rebasing, then you can resolve it locally and the history stays clean ;)

andersfylling commented:
I added the branch protection after the accident took place.

troiganto commented:
Part of the issue was also that, although the branch protection disallows force-pushing, it still allowed me to accidentally merge changes I didn't want to merge. In any case, I've modified my checkout to disable direct pushes to this repo, so this shouldn't happen again. Thank you again for your help!

andersfylling commented:
No, I added the protection after you accidentally merged.
1. When enabled, all commits must be made to a non-protected branch and submitted via a pull request with the required number of approving reviews and no changes requested before it can be merged into master. 
2. Choose which status checks must pass before branches can be merged into master. When enabled, commits must first be pushed to another branch, then merged or pushed directly to master after status checks have passed. 

so you cannot merge directly to master anymore. has to be a PR now, from how I read it.
